### ISSUE TITLE
Error `Cannot find name 'require' ` always shown while developing with webpack

### ISSUE THREAD

**Issue by leftstick** [TypeStrong/atom-typescript#829]
Let's say i have a piece of code as following:

```javascript
require.ensure(['less/main.less', 'splash-screen', './main'], function(require) {
    require('less/main.less');
    var App = require('./main').default;
    (new App()).run();
});
```

And my `tsconfig.json` is below:

```javascript
{
  "compilerOptions": {
    "target": "es5",
    "module": "es6",
    "declaration": false,
    "noImplicitAny": false,
    "removeComments": false,
    "noLib": false
  },
  "buildOnSave": false,
  "compileOnSave": false,
  "filesGlob": [
    "./js/**/*.ts",
    "./etc/config.ts"
  ],
  "formatCodeOptions": {
    "indentSize": 4,
    "tabSize": 4,
    "newLineCharacter": "\r\n",
    "convertTabsToSpaces": true,
    "insertSpaceAfterCommaDelimiter": true,
    "insertSpaceAfterSemicolonInForStatements": true,
    "insertSpaceBeforeAndAfterBinaryOperators": true,
    "insertSpaceAfterKeywordsInControlFlowStatements": true,
    "insertSpaceAfterFunctionKeywordForAnonymousFunctions": false,
    "insertSpaceAfterOpeningAndBeforeClosingNonemptyParenthesis": false,
    "placeOpenBraceOnNewLineForFunctions": false,
    "placeOpenBraceOnNewLineForControlBlocks": false
  },
  "files": [
    "./js/index.ts",
    "./etc/config.ts"
  ],
  "atom": {
    "rewriteTsconfig": true,
    "formatOnSave": true
  }
}
```

`Atom` always display errors `Cannot find name 'require'`, the `require` should be resolved by `webpack` while compiling. It is possible to have `atom-typescript`  plugin works well with `webpack`?

**Comment by basarat** [comment]
>  It is possible to have atom-typescript plugin works well with webpack

This is a general TypeScript question. You need to have `require` defined for TypeScript to know. The definition can be found here : https://github.com/TypeStrong/ts-loader#loading-other-resources-and-code-splitting

Feel free to continue discussion if anything is unclear :rose:

**Comment by leftstick** [comment]
@basarat Thanks for your quick answer. It saved me

**Comment by basarat** [comment]
no worries :rose:

**Comment by andry-tino** [comment]
I expected the definition to be available in commonjs website or in definitelytyped...

**Comment by aconfee** [comment]
I'm confused as well. It seems like this should definitely be build in. Do I really need to include that declaration? Where is the best place to include it?

I didn't see this extra step anywhere in this documentation: https://angular.io/docs/ts/latest/guide/webpack.html.

**Comment by aconfee** [comment]
Found the solution in the next Google search result :P My typings had not installed correctly. I installed 'npm install typings -g' and ran 'typings install' then recompiled and everything worked perfectly. 

Hopefully this is helpful to somebody later down the road.

**Comment by dcsan** [comment]
I always have this hassle with every node/typescript project starting up. 
I thought that 

`typings install --global --save dt~node`

should get TS & node to work together but something fragile seems to have broken.

any other clarity on how to fix this for ES6 compile target?

this is now outdated, correct:

     /// <reference path="typings/tsd.d.ts" />

(there's only a JSON file now anyway)

I'm using these compile options and vanilla `tsc`

```
{
    "compilerOptions": {
        "sourceMap": true,
        "target": "es6",
        "jsx": "react",
        "module": "commonjs",
        "moduleResolution": "node",
        "isolatedModules": false,
        "experimentalDecorators": true,
        "emitDecoratorMetadata": true,
        "declaration": false,
        "noImplicitAny": false,
        "removeComments": true,
        "noLib": false,
        "preserveConstEnums": true,
        "suppressImplicitAnyIndexErrors": true
    },
```

**Comment by googol7** [comment]
I don‘t use [typings](https://github.com/typings/typings) or [tsd](https://github.com/DefinitelyTyped/tsd) anymore. I install all typings like this: `npm install --save @types/node`

Therefore I need to include those typings in my index.ts with:
`///<reference path="../node_modules/@types/node/index.d.ts" />`

Then the error `Cannot find name 'require'` disappears as `require` is being defined like this:
`declare var require: NodeRequire;`

**Comment by wyqydsyq** [comment]
This actually seems to be an issue with the `atom-typescript` plugin in that it seems to not read the definitions from `@types/node` automatically.

My project compiles fine with `tsc` so clearly the Typescript compiler is discovering my ambient definitions meaning this is an issue with the Atom plugin directly and not a general Typescript question. Yet Atom complains about `Cannot find name 'require` unless I add the `<reference ... />` annotation to my entry file.

@basarat please reopen this issue for investigation.

**Comment by guncha** [comment]
Since people are still running into this, the recommended way to get Node types is:

1) Install using `yarn add @types/node` or `npm install @types/node --save-dev`
2) Add "node" to the "types" option in your tsconfig.json, so it looks like `"types": ["node"]`. This makes them available in any file in your project and you don't have to use reference comments.

More information: https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#types-typeroots-and-types

**Comment by dcsan** [comment]
this setup hassle is the major reason i won't use typescript for opensource projects. It seems crazy that unless you know that incantation above, node and ts don't work together... 

how about a default situation where nodeJS isn't broken?

**Comment by Saylux** [comment]
Ok, I tried everyone's suggestions and nothing helped. I was basically at a small hello world stage. I tried some random things and it fixed it though. I'm using vscode out of powershell on windows btw and was getting this in the console. what fixed it is I changed node to ts-node in tsconfig.json

`        "types": ["ts-node"],
        "typeRoots": ["node_modules/@types"]`

and then I also did
`npm install node -g`
`npm install ts-node -g`
`npm install typescript -g`

the last one was because I got a complaint that it couldn't see typescript.

I don't know much and am new to typescript and node and javascript, but thought I'd share my solution since the others didn't work even though I had the same issue
